bookkeeper-server/src/main/java/org/apache/bookkeeper/client/UpdateLedgerOp.java

```diff
@@ -81,174 +74,100 @@ public UpdateLedgerOp(final BookKeeper bkc, final BookKeeperAdmin admin) {
-                                updatedLedgerCnt, outstandings, syncObj, progressable);
-                        Futures.addCallback(readCb.getFutureListener(), updateLedgerCb);
-
-                        issuedLedgerCnt.incrementAndGet();
-                        if (limit != Integer.MIN_VALUE && issuedLedgerCnt.get() >= limit || !ledgerItr.hasNext()) {
-                            stop.set(true);
+        final CompletableFuture<Void> finalPromise = new CompletableFuture<>();
+        final Set<CompletableFuture<?>> outstanding =
+            Collections.newSetFromMap(new ConcurrentHashMap<CompletableFuture<?>, Boolean>());
+        final RateLimiter throttler = RateLimiter.create(rate);
+        final Iterator<Long> ledgerItr = admin.listLedgers().iterator();
+
+        // iterate through all the ledgers
+        while (ledgerItr.hasNext() && !finalPromise.isDone()
+               && (limit == Integer.MIN_VALUE || issuedLedgerCnt.get() < limit)) {
+            // throttler to control updates per second
+            throttler.acquire();
+
+            final long ledgerId = ledgerItr.next();
+            issuedLedgerCnt.incrementAndGet();
+
+            GenericCallbackFuture<LedgerMetadata> readPromise = new GenericCallbackFuture<>();
+            lm.readLedgerMetadata(ledgerId, readPromise);
+            CompletableFuture<LedgerMetadata> writePromise = readPromise.thenCompose((readMetadata) -> {
+                    AtomicReference<LedgerMetadata> ref = new AtomicReference<>(readMetadata);
+                    return new MetadataUpdateLoop(
+                            lm, ledgerId,
+                            ref::get,
+                            (metadata) -> {
+                                return metadata.getEnsembles().values().stream()
+                                    .flatMap(Collection::stream)
+                                    .filter(b -> b.equals(oldBookieId))
+                                    .count() > 0;
+                            },
+                            (metadata) -> {
+                                return replaceBookieInEnsembles(metadata, oldBookieId, newBookieId);
+                            },
+                            ref::compareAndSet).run();
+                });
+
+            outstanding.add(writePromise);
+            writePromise.whenComplete((metadata, ex) -> {
+                        if (ex != null
+                            && !(ex instanceof BKException.BKNoSuchLedgerExistsException)) {
+                            String error = String.format("Failed to update ledger metadata %s, replacing %s with %s",
+                                                         ledgerId, oldBookieId, newBookieId);
+                            LOG.error(error, ex);
+                            finalPromise.completeExceptionally(new IOException(error, ex));
+                        } else {
+                            LOG.info("Updated ledger {} metadata, replacing {} with {}",
```

Comment:
This line seems to be new. I think it changes the existing logging behavior. I would prefer converting it to DEBUG, since it is a refactor, better to keep behaviors unchanged